lib/LongVectorLoweringPass.cpp

```diff
@@ -1100,14 +1115,30 @@ clspv::LongVectorLoweringPass::visitInsertElementInst(InsertElementInst &I) {
          ScalarElement->getType()->isFloatingPointTy());
 
   ConstantInt *CI = dyn_cast<ConstantInt>(I.getOperand(2));
-  // We'd have to lower this to a store-store-load sequence.
-  assert(CI && "Dynamic indices not supported yet");
-  unsigned Index = CI->getZExtValue();
 
   IRBuilder<> B(&I);
-  auto *V = B.CreateInsertValue(EquivalentValue, ScalarElement, {Index});
-  registerReplacement(I, *V);
-  return V;
+  if (CI) {
+    unsigned Index = CI->getZExtValue();
+
+    auto *V = B.CreateInsertValue(EquivalentValue, ScalarElement, {Index});
+    registerReplacement(I, *V);
+    return V;
+  }
+
+  // Extract array pointer
+  Value *PtrOperand = nullptr;
+  if (auto *Load = dyn_cast<LoadInst>(EquivalentValue)) {
+    PtrOperand = Load->getPointerOperand();
+  } else {
+    PtrOperand = B.CreateAlloca(EquivalentValue->getType());
```

Comment:
This always makes me a bit nervous. Is it not a requirement that allocas are at the entry block of the function?

Reply:
Fixed, sorry for that.